### ISSUE TITLE
Saving to both managedObjectContext and mainThreadManagedObjectContext

### ISSUE THREAD

**Issue by vondiplo** [robbiehanson/XMPPFramework#564]
Hello,

In my app I've created my own custom module to handle roster information, save them as my own datatype and allow expending of the built in XMPPUserCoreDataStorageObject. 

My module works asynchronously on its own thread and builds up the user object.
When I want to fetch users I simplify this and fetch them directly from mainThreadManagedObjectContext. Now assume I wish to rename a user. I fetch it from the main thread context, and then changed a the value of the nickname property. In the meanwhile the user itself has changed on server (perhaps changed his nickname) and I receive this information via roster update. 

Now to my question, I have two different versions of the same instance, each on a different context and in a different thread. How do I merge them? Merely setting MergePolicy doesn't have an influence. Moreover, what context do I save to the DB? the mainThreadManagedObjectContext or the the thread dependent? 

Thanks

**Comment by ObjColumnist** [comment]
The mainThreadManagedObjectContext is designed to be read only, so you should make any changes in managedObjectContext.

Im not sure about your implementation (so it might not be an issue) but the provided XMPPRoster Implementation doesn't update users, it deletes and recreates them each time the roster is fetched.

**Comment by vondiplo** [comment]
Yes I understand that, but I wish to create a local database of rosters (which I extend) this for example keeps a populated list even if connection has not been established with server yet. 

So if I understand correctly, I should never call save on mainThreadManagedObjectContext ?

**Comment by ObjColumnist** [comment]
You can set properties on Managed Objects from the main thread, but because they are never saved you can lose these changes at any point ... so in short you shouldn't.

To make changes that are persisted you would need to add a method to you roster class that make changes using the internal MOC

**Comment by vondiplo** [comment]
Based on example used by the xmppframework, I notice that none of the object class has any reference to NSManagedObjectContext. Instead it seems as though the pattern is to hold a storage class for each NSManagedObject (my custom roster for example).

How would I go about updating my entities then? Directly through the storage classes or somehow through the entity itself? If I understand correctly, I shouldn't just go around modifying properties of objects returned via mainThreadManagedObjectContext, I should manage the change somehow.

Thank you very much for your patience and help

**Comment by vitalyster** [comment]
@ObjColumnist 
> it deletes and recreates them each time the roster is fetched.
But what is the point to use coredata in this case? I think it should be fixed

**Comment by ObjColumnist** [comment]
@vondiplo take a look at how handlePresence:xmppStream: works, but instead of passing in a presence pass in you custom value

**Comment by ObjColumnist** [comment]
@vitalyster it's more of cache than a persistence solution, so you can efficiently query it at runtime. This is mainly down to the fact that most servers don't support versioning so it just deletes the old roster and recreates it.

**Comment by vondiplo** [comment]
@ObjColumnist then if I understand correctly, all modification to instance should be done only via the storage class? I shouldn't modify the instance directly. Best add some public method that will update the property on the private managedobjectcontext. Did I understand correctly?

**Comment by vitalyster** [comment]
@ObjColumnist and what is the difference with XMPPRosterMemoryStorage?
Now there are many new extensions - versioning, session management, archive management, etc. What is the point to keep incomplete roster storage implementation?

**Comment by ObjColumnist** [comment]
@vondiplo yep 😊

**Comment by ObjColumnist** [comment]
@vitalyster CoreData is smart enough to discard values from memory and fetch them again when needed, plus you get efficient querying. You could argue that nobody has a roster big enough for that to be an issue mind.

**Comment by vitalyster** [comment]
@ObjColumnist but issue is "XMPPRosterCoreDataStorage delete roster each time", and every XMPPFramework user must implement custom storage, when we can fix it in here.

**Comment by ObjColumnist** [comment]
I agree with you in principle :smile: , but if you are not doing anything custom then you are not going to lose any data. You just don't have a roster for the brief period where the IQ is being parsed.
